chore: readme update for esm support to aws lambda (#1193)
instana/nodejs: packages/aws-lambda-auto-wrap/esm/README.md

> ## ES Module Support for AWS Lambda

Reviewer: ```suggestion
## ES Module Support for AWS Lambda Layer
```

AWS Lambda began supporting ES modules with Node v14. Learn more on the [AWS Lambda ES modules and top-level await with Node.js 14](https://aws.amazon.com/about-aws/whats-new/2022/01/aws-lambda-es-modules-top-level-await-node-js-14/) page.

- **Current Issues (July 2024)**:
  - Defining an ES module as a handler in a Lambda layer is not supported at present. For updates, refer to AWS support case ID 172069954500883.

- **Current Workaround**:
  - We utilize the handler as a CommonJS module with dynamic imports.
  - Transitioning the handler to pure ES modules does not provide immediate benefits, but we are monitoring AWS updates closely.

> - **Roadmap for Improvement**:

Reviewer: ```suggestion
- **Roadmap for Improvements**:
```

>   - We've initiated an AWS support ticket (172069954500883) and engaged with the AWS SDK community to advocate for ES module support in Lambda layers.

Reviewer: Could you please add the discussion link from the AWS SDK Github Repo?

Author: added.

>   - Future plans for version 4.x include removing the `aws-lambda-auto-wrap` npm package and enhancing our deployment process by manually integrating necessary files during the layer publishing script.

Reviewer: ```suggestion
  - In 4.x we will remove the `aws-lambda-auto-wrap` npm package, because this is an internal package. It got published by mistake. We copy the necessary files during the layer publishing script into the layer.
```

>   - Considering discontinuing support for Node.js version 14, which have reached the end of their LTS (Long-Term Support) lifecycle.

Reviewer: ```suggestion
  - Considering discontinuing support for Node.js version 14, which has reached the end of their LTS (Long-Term Support) lifecycle.
```

We are affected by the second bug. When AWS fixes the underlying problem in their AWS runtime, we are able to transform the new esm handler into a real ES module.

For the time being, we are shipping the ES handler as a CommonJS module using dynamic imports as a workaround for Node.js 14 and 16, as AWS has yet to resolve the issue for these versions, which are no longer LTS.
see the comment https://github.com/aws/aws-sdk-js-v3/issues/3230#issuecomment-1561973247

> The Node.js 18.x runtime introduces support for ES module resolution using NODE_PATH. For more information, refer to [Node.js 18.x runtime now available in AWS Lambda](https://aws.amazon.com/blogs/compute/node-js-18-x-runtime-now-available-in-aws-lambda/). This is the recommended workaround.

Reviewer: I am not able to follow the connection between the [ES module resolution using NODE_PATH](https://aws.amazon.com/blogs/compute/node-js-18-x-runtime-now-available-in-aws-lambda/) and the Lambda layer  🤔 Could you please explain how that fixes the layer?

Refs:

> 2. You cannot define an ES module as handler, which is located in a layer. See AWS support case ID 11226031451.

> We are affected by the second bug. When AWS fixes the underlying problem in their AWS runtime, we are able to transform the new esm handler into a real ES module.

Did you test deploying our layer with the ESM handler with native ESM code? What was the result?
If yes and it was successful, what do we need to change?

Author: I haven't able to test this, will do it and update the status

Author: I'm getting an error like this with transftering the layer handler to pure ESM  "Error: Cannot find module 'instana-aws-lambda-auto-wrap-esm'\nRequire stack:\n- /var/runtime/index.mjs'

Reviewer: Could you please share the steps you did?
If the esm handler would be a native handler, how would the handler look like and would we benefit from it?

Author: I will add details in the box note.

Reviewer: Is this done? Could you send me the link in Slack? Thanks <3
